### DISCUSSION TITLE
visualize ThreadDataLoader and GPU transform in Nsight Systems

### DISCUSSION THREAD

**Discussion by hw-ju** [Project-MONAI/tutorials#1173] [Q&A]
Hi!
The profiling section in the [fast_training_tutorial](https://github.com/ProjectMONAI/tutorials/blob/main/acceleration/fast_training_tutorial.ipynb) is very useful! 
I'm trying to visualize the GPU transform and ThreadDataLoader in the fast training in NsighT Systems UI. Clicking on the range for GPU transform "RandCrop" doesn't correlate with any kernels on the GPU row, please see the two screenshots below:
![Screenshot from 2023-01-16 10-33-23](https://user-images.githubusercontent.com/90806807/212748979-4f9358ab-b646-4454-a813-ac427671f81d.png)
![Screenshot from 2023-01-16 10-33-50](https://user-images.githubusercontent.com/90806807/212748989-e2cc11cf-3f7d-4989-a6f7-f78688ad2739.png)

I'm wondering if the CUDA API call "CatArrayBatchedCopy" right after the range "RandCrop" is the one responsible for launching the GPU transform?
![Screenshot from 2023-01-16 10-47-38](https://user-images.githubusercontent.com/90806807/212749161-93ca6f3b-4645-4218-83ff-70fef3a81f67.png)

Another quick question, in fast training, the "dataload" range doesn't project to the GPU row, please see the highlighted area in the screenshot below. Is that because nothing is happening on the GPU during that range (i.e., no host-to-device data movement as in regular PyTorch training)?
![Screenshot from 2023-01-16 11-00-19](https://user-images.githubusercontent.com/90806807/212749685-43c26279-f4e3-4237-a39a-1f052f8ad2ec.png)

**Comment by Nic-Ma** [comment]
Hi @KumoLiu ,

Could you please help share some comments about this question? I think you are also using the Nsight system recently.

Thanks in advance.

**Comment by KumoLiu** [comment]
I tried to do some comparative experiments and the results are as follows. The height of the blue bars represents utilization rate.
| __Transform__ |  RandCropByPosNegLabeld   | RandAffined  | RandFlipd |
|  ----  | ----  |  ----  | ----  |
| __GPU__  | ~4ms | ~7.5ms | ~260μs
| __CPU__  | ~4ms | ~52.5ms | ~1.5ms
|__GPU Usage__|![Screen Shot 2023-01-18 at 16 53 26](https://user-images.githubusercontent.com/55491388/213126510-4b7611b3-de30-49dd-9f54-4242c5431723.png)| ![Screen Shot 2023-01-18 at 16 54 06](https://user-images.githubusercontent.com/55491388/213126673-71347f17-ad3a-43ea-916a-24225184f77d.png)|![Screen Shot 2023-01-18 at 16 54 32](https://user-images.githubusercontent.com/55491388/213126753-e4a95f24-e14f-4444-9e86-bc892952fa9a.png)

From the table, we could see that `RandAffined` has the relatively most significant speed increase when it is moved to the GPU. I guess the reason for this is that affine has more matrix operations, which is more conducive to parallel computing on the GPU. In comparison, the `RandCropByPosNegLabeld` is merely a slicing operation which may not be conducive to using the parallel computing of GPU to speed up, but the use of `EnsureType` also reduces the transfer time between different devices, and the overall speed of training is improved (7s/epoch vs 9s/epoch). I'm not sure for the reason why the GPU utilization will improve after crop in each iteration.
https://github.com/Project-MONAI/MONAI/blob/79b8b0ca2a43586e754a97f0e36ae8e5e551b94b/monai/transforms/croppad/array.py#L1107

Do you have any other insight on this？ @Nic-Ma @wyli @yiheng-wang-nv
Also thanks for the guidance from @yiheng-wang-nv!
